File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        stack<TreeNode*> stack;
        TreeNode* node = root; 
        while(stack.size() || node){
            if(node){
                stack.push(node);
                node=node->left;
            }
            else{
                TreeNode* top = stack.top();
                stack.pop();
                k--;
                if(!k) return top->val;
                node = top->right;
            }
        }
        return -1;
    }
};

File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.cnt = 0 
        self.res = None 

    def dfs(self,node):
        if node is None: 
            return 
        
        self.dfs(node.left)
        self.cnt-=1 
        if self.cnt==0: 
            self.res=node

        self.dfs(node.right)

    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        self.cnt = k 
        self.dfs(root,)
        return self.res.val 
        